### DISCUSSION TITLE
Strict comparaison in project

### DISCUSSION THREAD

**Discussion by ColonelMoutarde** [FreshRSS/FreshRSS#4445] [Ideas]
Hello,
I made some remarks recently about the use of non-strict comparisons in the project. It was written that some non-strict comparisons were chosen carefully. I understand this point of view.

Let me shed some light on the difference between strict and non-strict which is important.
https://www.php.net/manual/en/types.comparisons.php
Here is the table of non-strict comparisons
![Screenshot 2022-07-13 at 16-54-30 PHP PHP type comparison tables - Manual](https://user-images.githubusercontent.com/4697568/178764794-88a46ebd-7d98-4119-9b7b-e2a924b4e05a.png)
It is particularly difficult to read and the results are often unexpected. 
it is also necessary to take into account that the behavior changes between the version 7.4 and the version 8 of php (the asterisk in bottom on the left).
![Screenshot 2022-07-13 at 16-58-33 PHP PHP type comparison tables - Manual](https://user-images.githubusercontent.com/4697568/178765637-672b8722-a5a9-4a23-bb22-cb882a07a508.png)

it is also necessary to take into account that the behavior changes between the version 7.4 and the version 8 of php (the asterisk in bottom on the right).

What is your opinion ?

**Comment by Alkarex** [comment]
Hello,
My opinion is to use what makes sense for a given case, and refrain from comparing values of very different types (so avoid comparisons between e.g. an array and a string).

In particular, it is often useful to test whether something is *truthy* or *falsy* and for that a non-strict comparison is used, either in an implicit form `if ($something)` or explicitly with the most expected type such as `if ($something != '')` 

Likewise, for boolean expressions, I would rather keep to a short syntax `if (fine())` and not `if (fine() === true)`

We also often have integers as string for which `if ($a == 3)` is the right approach, at least for the code paths where upstream types are not better enforced.

Furthermore, I also use and like git logs (e.g. to help remembering why a line is how it is) and I am not found of making near-cosmetic / minor changes that are not a part of a new feature / bug fix, especially not when fragmented in multiple PRs (due to processing / testing overhead). Example https://github.com/FreshRSS/FreshRSS/pull/4438

I am already paying attention to those things and make appropriate changes or ask for pull request changes when the comparison should be improved, e.g. https://github.com/FreshRSS/FreshRSS/pull/4101#discussion_r914591078

I am more keen of making such changes in code sections where types are better enforced, e.g. as part of our PHPStan improvements https://github.com/FreshRSS/FreshRSS/issues/4112 and especially when it is required for an automated test.

In any case, blindly changing comparisons is not OK, as it will introduce bugs, and it must always be evaluated and tested. Example https://github.com/FreshRSS/FreshRSS/pull/4439#discussion_r919513316

All that being said, if you spot any instance of comparisons, which results have changed in PHP 8, a pull request for that would be very welcome :-)
Likewise, fixes for comparisons between types that should not be directly compared (e.g. arrays with string, objects with numbers...) are appreciated.

I hope this clarifies a few things :-)

**Reply by ColonelMoutarde** [reply]
Thank you very much for this answer, I will try to take it into account ;)
